source/particles/utilities.cc

```diff
@@ -0,0 +1,221 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2018 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/lac/generic_linear_algebra.h>
+
+#include <deal.II/particles/utilities.h>
+
+
+DEAL_II_NAMESPACE_OPEN
+
+
+namespace Particles
+{
+  namespace Utilities
+  {
+    template <int dim, int spacedim, typename Sparsity, typename number>
+    void
+    create_interpolation_sparsity_pattern(
+      const DoFHandler<dim, spacedim> &                space_dh,
+      const Particles::ParticleHandler<dim, spacedim> &particle_handler,
+      Sparsity &                                       sparsity,
+      const AffineConstraints<number> &                constraints,
+      const ComponentMask &                            space_comps)
+    {
+      if (particle_handler.n_locally_owned_particles() == 0)
+        return; // nothing to do here
+
+      const auto &tria     = space_dh.get_triangulation();
+      const auto &fe       = space_dh.get_fe();
+      auto        particle = particle_handler.begin();
+      const auto  max_particles_per_cell =
+        particle_handler.n_global_max_particles_per_cell();
+
+
+      // Take care of components
+      const ComponentMask comps =
+        (space_comps.size() == 0 ? ComponentMask(fe.n_components(), true) :
+                                   space_comps);
+      AssertDimension(comps.size(), fe.n_components());
+
+      const auto n_comps = comps.n_selected_components();
+
+      // Global to local indices
+      std::vector<unsigned int> space_gtl(fe.n_components(),
+                                          numbers::invalid_unsigned_int);
+      for (unsigned int i = 0, j = 0; i < space_gtl.size(); ++i)
+        if (comps[i])
+          space_gtl[i] = j++;
+
+      // [TODO]: when the add_entries_local_to_global below will implement
+      // the version with the dof_mask, this should be uncommented.
+      // // Construct a dof_mask, used to distribute entries to the sparsity
```

Comment:
Do you want to keep this TODO?

Reply:
Yes. This is here to remind us what to do when the corresponding function is implemented, instead of throwing an `ExcNotImplemented` exception.